### ISSUE TITLE
Filters (e.g. Gaussian blur) do not work

### ISSUE THREAD

**Issue by corliss** [vvvv/SVG#43]
SvgGaussianBlur has a bunch of code for computing the gaussian blur of an image, but none of it seems to actually be called. Am I missing something? Cheers.

**Comment by mrbean-bremen** [comment]
Currently, none of the W3C filter examples seems to work.
The filter classes are created dynamically after a class lookup using reflection - though there is no direct call in the code, they are applied, so that is not the reason why they are not working.

**Comment by AnReZa** [comment]
I'm also affected by this issue. I'm using a matrix transform filter to sharpen parts of the image. Instead, the whole image appears black.

**Comment by mrbean-bremen** [comment]
Yes, I didn't really look into this until now - I hoped that someone with a better understanding would have a look (like the original developers...). Maybe I will have a go at this sometime later, but this doesn't look like an easy fix.

**Comment by AnReZa** [comment]
Yes, just by the sheer fact, that this ticket is open for years now, I guess, that this isn't trivial to fix. But thank you anyways for your help!

**Comment by mrbean-bremen** [comment]
>  I'm using a matrix transform filter to sharpen parts of the image. Instead, the whole image appears black.

Not sure if this is your problem, but if there are floats in your matrix, and you use a (non-English) locale where the decimal point is a comma, there has been a bug in reading the float values (fixed a few days ago).
Anyway, I'm going to have a look at the filter stuff, and if you have a concrete example image of a filter not working, I may have a look at that first.

**Comment by AnReZa** [comment]
I guess, that this shouldn't be the case here. I've attached one example SVG for you. You can replace the linked picture with anything you like. The image should show a partially sharpened area in the lower right corner.
[Test.zip](https://github.com/vvvv/SVG/files/3176184/Test.zip)

**Comment by mrbean-bremen** [comment]
Thanks, I will have a look later!

**Comment by mrbean-bremen** [comment]
Ok, I can see the original image, but no filter applied.

**Comment by mrbean-bremen** [comment]
...which is no surprise after checking the implementation: the only implemented filters are feColourMatrix, feGaussianBlur (partly), feMerge and feOffset.

**Comment by mrbean-bremen** [comment]
These are exactly the ones that are used in one of the example images - `__issue-083-01.svg`. `feGaussianBlur` works with one value for `stdDeviation`, but not with two, and the result image looks different from the one in the browser (less blurred). The other 3 (more basic ones) seem to work correctly.

For reference, the following filter primitives are not implemented:
`feBlend`, `feComponentTransfer`, `feComposite`, `feConvolveMatrix`, `feDiffuseLighting`, `feDisplacementMap`, `feFlood`, `feImage`, `feMorphology`, `feSpecularLighting`, `feTile` and `feTurbulence`.

**Comment by AnReZa** [comment]
Do you think, that it would be possible to implement `feConvolveMatrix`, `feComponentTransfer` and `feBlend` in a reasonable amount of time? Those are the ones used in my project. Maybe they can be adopted from the Inkscape source code or some other open source SVG rendering engine.

**Comment by mrbean-bremen** [comment]
Well, it depends on who will do the work, and what is a reasonable amount of time ;)
I started to look at the filter code, but I have no experience there (and not that much free time), so this may take some time, but I cannot say how much at the moment. Any help would be appreciated, of course!

**Comment by AnReZa** [comment]
Unfortunately I can't help you much. I'm horrible at maths and don't know the code at all.

**Comment by mrbean-bremen** [comment]
No problem, I just can't promise anything...

**Comment by megakraken** [comment]
Has there been any progress with this at all? It would be amazing if we could get more of the filters to work, especially the lighting filters such as `feSpecularLighting` and `feDiffuseLighting` are just such a powerful and widely used asset of SVG and it would be awesome if the library supported them.

**Comment by mrbean-bremen** [comment]
Unfortunately, no progress here. I had abandoned the PR I intialially started, it didn't get anywhere (I actually forgot the details), and so far nobody else has tried this.

**Comment by wieslawsoltes** [comment]
> Has there been any progress with this at all? It would be amazing if we could get more of the filters to work, especially the lighting filters such as `feSpecularLighting` and `feDiffuseLighting` are just such a powerful and widely used asset of SVG and it would be awesome if the library supported them.

You can try using https://github.com/wieslawsoltes/Svg.Skia

NuGet: https://www.nuget.org/packages/svg.skia

It's based on this library but requires SkiaSharp.

**Comment by mrbean-bremen** [comment]
@wieslawsoltes - is #590 still something to consider? It would be a big progress, an would also help this this kind of issues.

**Comment by wieslawsoltes** [comment]
> @wieslawsoltes - is #590 still something to consider? It would be a big progress, an would also help this this kind of issues.

@mrbean-bremen Yes, but not really possible to integrate with Svg.Skia at this point. I would still prefer splitting but preferably via partial classes, not really sure if that would work. I have some ideas, but not much time on hand.

**Comment by mrbean-bremen** [comment]
Thanks - good to hear that this is still an option! Time constraints are an issue most of us have...

**Comment by megakraken** [comment]
> > Has there been any progress with this at all? It would be amazing if we could get more of the filters to work, especially the lighting filters such as `feSpecularLighting` and `feDiffuseLighting` are just such a powerful and widely used asset of SVG and it would be awesome if the library supported them.
> 
> You can try using https://github.com/wieslawsoltes/Svg.Skia
> 
> NuGet: https://www.nuget.org/packages/svg.skia
> 
> It's based on this library but requires SkiaSharp.

interesting, I will take a look, thank you!

**Comment by megakraken** [comment]
it doesn't look _that_ complicated to implement these filters. maybe i'll try to mess around with it a bit...
